src/mds/MDSRank.cc

```diff
@@ -3054,11 +3054,77 @@ bool MDSRankDispatcher::handle_command(
 
     damage_table.erase(id);
     return true;
+  } else if (prefix == "drop cache") {
+    command_drop_cache(ds);
+    return true;
   } else {
     return false;
   }
 }
 
+void MDSRank::command_drop_cache(stringstream *ds)
+{
+  if (not is_active()) {
+    dout(5) << __func__ << ": MDS not active, no-op" << dendl;
+    return;
+  }
+
+  if (not mdcache->trim(UINT64_MAX))
+    return;
+
+  server->recall_client_state(1, true);
```

Comment:
pass a pointer to

`MDSGatherBuilder gather(g_ceph_context);`

to `server->recall_client_state` which adds to the gather:

https://github.com/ceph/ceph/pull/21566/files#diff-eecf840e09d7481f7b8f415789852c7dR1177

then you can do a `gather.activate()` in `_wait_for_client_recall_state_acks` instead of the cpu-intensive loop.

Reply:
I can't figure out how to replace the busy/wait loop by `gather.activate()`. Also, calling `gather.activate()` in `_wait_for_client_recall_state_acks` leads to a crash due to the assert in `gather.activate()` [1]

And would these changes mean no need timeout? If so, the option `mds_session_recall_ack_timeout` will be redundant.

[1] https://github.com/ceph/ceph/blob/master/src/include/Context.h#L476